deepface/modules/recognition.py

```diff
@@ -120,35 +123,59 @@ def find(
     with open(datastore_path, "rb") as f:
         representations = pickle.load(f)
 
-    # Check if the representations are out-of-date
-    if len(representations) > 0:
-        if len(representations[0]) != len(df_cols):
+    # check each item of representations list has required keys
+    for i, current_representation in enumerate(representations):
+        missing_keys = list(set(df_cols) - set(current_representation.keys()))
+        if len(missing_keys) > 0:
             raise ValueError(
-                f"Seems existing {datastore_path} is out-of-the-date."
-                "Please delete it and re-run."
+                f"{i}-th item does not have some required keys - {missing_keys}."
+                f"Consider to delete {datastore_path}"
             )
-        pickled_images = [representation[0] for representation in representations]
-    else:
-        pickled_images = []
+
+    # embedded images
+    pickled_images = [representation["identity"] for representation in representations]
 
     # Get the list of images on storage
     storage_images = __list_images(path=db_path)
 
+    if len(storage_images) == 0:
+        raise ValueError(f"No item found in {db_path}")
+
     # Enforce data consistency amongst on disk images and pickle file
     must_save_pickle = False
-    new_images = list(set(storage_images) - set(pickled_images)) # images added to storage
-    old_images = list(set(pickled_images) - set(storage_images)) # images removed from storage
+    new_images = list(set(storage_images) - set(pickled_images))  # images added to storage
+    old_images = list(set(pickled_images) - set(storage_images))  # images removed from storage
+
+    # detect replaced images
+    replaced_images = []
+    for current_representation in representations:
+        identity = current_representation["identity"]
+        if identity in old_images:
+            continue
+        alpha_hash = current_representation["hash"]
+        beta_hash = package_utils.find_hash_of_file(identity)
+        if alpha_hash != beta_hash:
+            logger.debug(f"Even though {identity} represented before, it's replaced later.")
+            replaced_images.append(identity)
```

Comment:
This is a bit expensive.
You compute the hash of a file here which is eventually discarded.
Then the file with unmatched hash is added to the list of new files which are passed to `_find_bulk_embeddings` where the hash of the file is recomputed once again.

Reply:
I suggest, instead of hashing the whole content of the file, to hash only most immediate (and lighter) properties: name, creation timestamp, last modification timestamp

Reply:
Liked that approach and convinced. Will do it in another PR soon.